js/ClockCtrl.js

```diff
@@ -1,348 +1,272 @@
-        return map;
-    }, {});
-}
-
-angular.module('Clock',['ngStorage'])
-    .filter('time',function() {
-        function pad(str) {
-            str = ''+str;
-            if (str.length<2) {
-                str = '0'+str;
-            }
-            return str;
-        }
-        return function(msec,hundreds) {
-            var secs = Math.floor(msec/1000);
-            var hsec = Math.floor((msec/100)) % 10;
-            var sec = secs % 60;
-            var min = Math.floor((secs-sec)/60);
-            var str = min+':'+pad(sec);
-            if (hundreds) {
-                str += '.'+hsec;
-            }
-            return str;
-        };
-    })
-    .factory('$audio',[
-        function() {
-            function init(file,cb) {
-                new Audio5js({
-                    swf_path: 'swf/audio5js.swf',
-                    ready: function () {
-                        this.load(file);
-                        this.reset = function() {
-                            this.pause();
-                            this.load(file);
-                        };
-                        cb(this);
-                    }
-                });
-            }
-
-            return {
-                init: init
-            };
-        }
-    ])
-
-    .controller('ClockCtrl',[
-        '$scope','$timeout','$audio','$window','$localStorage',
-        function($scope,$timeout,$audio,$window,$localStorage) {
-            //initial values
-            $audio.init('mp3/lossetrack-A +6.mp3',function(track) {
-                $scope.runTrack = track;
-            });
-            $audio.init('mp3/lossetrack-B.mp3',function(track) {
-                $scope.stopTrack = track;
-            });
-            //initialize config with the angular configuration
-            var urlConfig
-            try {
-                // console.log(params().state);
-                urlConfig = JSON.parse(params().state);
-            } catch(e) {
-                //no url Config
-            }
-
-            //config from localStorage, then url, then defaults from config
-            $scope.$storage = $localStorage.$default({
-                config: urlConfig || clockConfig
-            });
-            var handlers = {};
-            $scope.bgColor = 'black';
-            $scope.state = 'stopped';
-            $scope.time = $scope.$storage.config.seconds * 1000;
-            $scope.armTime = $scope.$storage.config.seconds * 1;
-            $scope.tenths = false;
-            $scope.size = 340;
-
-            $scope.pos = {
-                x: 0,
-                y: 0
-            };
+angular.module('Clock').controller('ClockCtrl',[
+    '$scope','$timeout','$window','$config','$tracks',
+    function($scope,$timeout,$window,$config,$tracks) {
+        $config.init().then(function(config) {
+            $scope.config = config;
+            $scope.time = config.seconds * 1000;
+            $scope.armTime = config.seconds * 1;
+            $scope.$apply();

```

Comment:
This is bad practice $apply should only be used when an action is performed outside of angular context. For example, on a native browser event.

In this case, a promise is resolved somewhere in the future, at which point, there is no angular digest going on. Adding $apply here does solve the problem, but it is the wrong way to do it. Rather, it should be handled in the service itself, so that the user of that service does not need to think about having to apply.

In general, when you for example create a directive that responds to events, you apply there.

In this case, the correct way to do it is to wrap the promise in $q, using $q.when. A digest is then automatically triggered when the promise resolves.